Lib/functools.py

```diff
@@ -192,6 +192,222 @@ def total_ordering(cls):
+        will be the union of all dependencies passed in.
+
+        It is possible to add a node with no dependencies (*predecessors* is not provided)
+        as well as provide a dependency twice. If a node that has not been provided before
+        is included among *predecessors* it will be automatically added to the graph with
+        no predecessors of its own.
+
+        Raises ValueError if called after "prepare".
+        """
+        if self.ready_nodes is not None:
+            raise ValueError("Nodes cannot be added after a call to prepare()")
+
+        # Create the node -> predecessor edges
+        nodeinfo = self._get_nodeinfo(node)
+        nodeinfo.npredecessors += len(predecessors)
+
+        # Create the predecessor -> node edges
+        for pred in predecessors:
+            pred_info = self._get_nodeinfo(pred)
+            pred_info.successors.append(node)
+
+    def prepare(self):
+        """Mark the graph as finished and check for cycles in the graph.
+
+        If any cycle is detected, "CycleError" will be raised, but "get_ready" can
+        still be used to obtain as many nodes as possible until cycles block more
+        progress. After a call to this function, the graph cannot be modified and
+        therefore no more nodes can be added using "add".
+        """
+        if self.ready_nodes is not None:
+            raise ValueError("cannot prepare() more than once")
+
+        self.ready_nodes = [i.node for i in self.node2info.values()
+                            if i.npredecessors == 0]
+        # ready_nodes is set before we look for cycles on purpose:
+        # if the user wants to catch the CycleError, that's fine,
+        # they can continue using the instance to grab as many
+        # nodes as possible before cycles block more progress
+        cycle = self._find_cycle()
+        if cycle:
+            raise CycleError(f"nodes are in a cycle", cycle)
+
+    def get_ready(self):
+        """Return a tuple of all the nodes that are ready.
+
+        Initially it returns all nodes with no predecessors and once those are marked
+        as processed by calling "done", further calls will return all new nodes that
+        have all their predecessors already processed until no more progress can be
+        made.
+
+        Raises ValueError if called without calling "prepare" previously.
+        """
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+
+        # Get the nodes that are ready and mark them
+        result = tuple(self.ready_nodes)
+        n2i = self.node2info
+        for node in result:
+            n2i[node].npredecessors = _NODE_OUT
+
+        # Clean the list of nodes that are ready and update
+        # the counter of nodes that we have returned.
+        self.ready_nodes.clear()
+        self.npassedout += len(result)
+
+        return result
+
+    def is_active(self):
```

Comment:
nit: personally for me ``has_more_nodes`` would read nicer; I get the idea here is that the algorithm is still active... but that feels to read worse if you don't think about the object a generator 🤔

Reply:
The advantage of `is_active()` is that it's so vague you'll read the docs :wink:.

It does _not_ mean "has more nodes", for example.  What it means is subtler than that:  it means there are still more nodes to pass out and/_or_ that there are  nodes that were passed out that haven't yet checked in (been marked `done()`).

It's quite possible, e.g., that after `is_active()` returns `True`, `get_ready()` returns an empty tuple.  In fact, every topsort will end that way, right before the last leaf node is passed to `done()`.